### PR TITLE
fix: unknown values from PlanResourceChange are not handled correctly

### LINKED ISSUE #1943: Rewrite objectFromCtyValue function
### What happened?

We need to rewrite the [objectFromCtyValue](https://github.com/pulumi/pulumi-terraform-bridge/blob/935ad83006358781cdd9b66b85a872e296c6c0d5/pkg/tfshim/sdk-v2/instance_state.go#L69-L86) function. We will need to implement our own transform function to handle converting the cty value to a JSON map.

Currently this function performs some type checking which we do not want it to do. This leads to issues like #1932 

### Example

This can be used to reproduce the error

```go
package main

import (
	"fmt"

	"github.com/pulumi/pulumi-aws/sdk/v6/go/aws/ec2"
	"github.com/pulumi/pulumi-aws/sdk/v6/go/aws/s3"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {

		bucket, err := s3.NewBucket(ctx, "my-bucket", nil)
		if err != nil {
			return err
		}

		template, err := ec2.NewLaunchTemplate(ctx, "zbuchheit-launch-template", &ec2.LaunchTemplateArgs{
			ImageId: pulumi.String("dummyImageIdBecauseItIsAPreview"),
			TagSpecifications: &ec2.LaunchTemplateTagSpecificationArray{
				&ec2.LaunchTemplateTagSpecificationArgs{
					ResourceType: pulumi.String("resourceType1"),
					Tags: pulumi.StringMap{
						"Name": pulumi.Sprintf("BucketArn %s", bucket.Arn),
					},
				},
				&ec2.LaunchTemplateTagSpecificationArgs{
					ResourceType: pulumi.String("resourceType2"),
					Tags: pulumi.StringMap{
						"Name": pulumi.String("Bucket Arn"), // If you make this an output also, the panic will go away
					},
				},
			},
		},
		)
		if err != nil {
			return err
		}

		ctx.Export("bucket", bucket)
		ctx.Export("template", template)
		return nil
	})
}
```

### Output of `pulumi about`

CLI          
Version      3.111.1
Go Version   go1.22.1
Go Compiler  gc

Plugins
NAME  VERSION
aws   6.33.1
go    unknown

Host     
OS       darwin
Version  14.4.1
Arch     arm64

This project is written in go: executable='/usr/local/go/bin/go' version='go version go1.22.0 darwin/arm64'

Current Stack: chall-pulumi-corp/pulumi-go-app/dev

TYPE                  URN
pulumi:pulumi:Stack   urn:pulumi:dev::pulumi-go-app::pulumi:pulumi:Stack::pulumi-go-app-dev
pulumi:providers:aws  urn:pulumi:dev::pulumi-go-app::pulumi:providers:aws::default
aws:iam/role:Role     urn:pulumi:dev::pulumi-go-app::aws:iam/role:Role::myrole

Found no pending operations associated with dev

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/chall-pulumi-corp
User           chall-pulumi-corp
Organizations  chall-pulumi-corp, pulumi
Token type     personal

Dependencies:
NAME                                 VERSION
github.com/pulumi/pulumi-aws/sdk/v6  v6.33.1
github.com/pulumi/pulumi/sdk/v3      v3.115.2

Pulumi locates its logs in /var/folders/3b/6mr1jkqx7r797ff75k27jfjc0000gn/T/ by default

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by corymhall**
When `PlanResourceChange` is enabled, the schema that is passed to `MakeTerraformOutputs` is a `cty.Value`. We then have to convert this into a JSON object as part of `objectFromCtyValue` before eventually converting to Pulumi values. We were relying on a couple of functions from other packages to do this conversion between `cty.Value` and a JSON object which led to issues like #1932.

For example, if we have a map value like this:
```json
{
  "tagSpecifications": [
    {
      "tags": {
        "Name": "04da6b54-80e4-46f7-96ec-b56ff0331ba9"
      }
    }
    {
      "tags": {
        "Name": "Bucket Arn"
      }
    }
  ]
}
```

The first `tags` map contains an `unknown` sentinel so it will be converted by `PlanResourceChange` into something like this:

```
{
  "tagSpecifications": [
    {
      "tags": UnknownMapValue
    }
    {
      "tags": {
        "Name": "Bucket Arn"
      }
    }
  ]
}
```

As part of the conversion we were using the `cty.Transform` function which performs some type checking to ensure that every item in the list is of the same type.

Since all we want from this function is to convert `cty.Value` to a JSON object, I've taken the underlying `cty.marshal` function and modified it to handle unknown values.

fixes #1943, re #1932

### CODE REVIEW

File: pkg/tfshim/sdk-v2/cty_json.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+		b.WriteString("null")
+		return nil
+	}
+
+	// This is the one difference between the hashicorp/go-cty mashal function
+	// This function does not correctly handle Unknown values. At this point it could be any unknown type,
+	// an unknown map, list, etc, but it is not possible to further recurse on an Unknown type
+	// so the best we can do is just convert it to a string sentinel
+	if !val.IsKnown() {
+		b.WriteString(fmt.Sprintf("%q", terraformUnknownVariableValue))
```
Comment (t0yv0) [inline pkg/tfshim/sdk-v2/cty_json.go:40]: Excellent.

**Review by t0yv0 — APPROVED** [review]
Thanks a lot for a quick fix here!

File: pkg/tfshim/sdk-v2/cty_json.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+	if val.IsNull() {
+		b.WriteString("null")
+		return nil
+	}
+
+	// This is the one difference between the hashicorp/go-cty mashal function
+	// This function does not correctly handle Unknown values. At this point it could be any unknown type,
+	// an unknown map, list, etc, but it is not possible to further recurse on an Unknown type
+	// so the best we can do is just convert it to a string sentinel
```
Comment (iwahbe) [inline pkg/tfshim/sdk-v2/cty_json.go:38]: Thanks for the comment here.

**Review by iwahbe — APPROVED** [review]

File: pkg/tfshim/sdk-v2/cty_json.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+	"encoding/json"
+	"fmt"
+	"sort"
+
+	"github.com/hashicorp/go-cty/cty"
+	cty_json "github.com/hashicorp/go-cty/cty/json"
+)
+
+// marshal takes a cty.Value and coverts it to a JSON string
+// This is a fork of the mashal function for the hashicorp/go-cty package
```
Comment (marco-m) [inline pkg/tfshim/sdk-v2/cty_json.go:14]: typos: coverts, mashal

File: pkg/tfshim/sdk-v2/cty_json.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+	if t == cty.DynamicPseudoType && val.Type() != cty.DynamicPseudoType {
+		return marshalDynamic(val, path, b)
+	}
+
+	if val.IsNull() {
+		b.WriteString("null")
+		return nil
+	}
+
+	// This is the one difference between the hashicorp/go-cty mashal function
```
Comment (VenelinMartinov) [inline pkg/tfshim/sdk-v2/cty_json.go:35]:
```suggestion
	// This is the one difference between the hashicorp/go-cty marshal function
```

**Review by VenelinMartinov — APPROVED** [review]

**Comment by t0yv0** [comment]
Thanks all for spot-checking the language! As a non-native English speaker I always dread these, a bit better since I installed JINX spell-checker in my Emacs, perhaps there's more powerful tools we can install in CI checks to help us get the veribage right?

**Comment by VenelinMartinov** [comment]
I have a spell-checker in my editor, it tends to be very error prone, since we have so many contractions/made up words. Not sure how we can adapt that to CI without a very large dictionary of pulumi-specific words

**Comment by VenelinMartinov** [comment]
Maybe we can run a spell-checker only against newly added code and commit to maintaining a list of word exceptions - I'd be happy to contribute what I have so far.

**Pull request merged**